imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -260,6 +266,127 @@ def set_spin_cycle(l1a_de: xr.Dataset, l1b_de: xr.Dataset) -> xr.Dataset:
+
+    The resulting array of spin start times will be equal to the length of the direct
+    events. If two direct events occurred in the same spin, then there will be repeating
+    spin start times.
+
+    Parameters
+    ----------
+    l1a_de : xr.Dataset
+        The L1A DE dataset.
+    l1b_de : xr.Dataset
+        The L1B DE dataset.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+    acq_end : xr.DataArray
+        The end acquisition times for each spin ASC.
+
+    Returns
+    -------
+    spin_start_time : xr.DataArray
+        The start time for the spin that each direct event is in.
+    """
+    met = l1a_de["met"].values
+    # Find the closest stop_acq for each shcoarse
+    closest_stop_acq_indices = np.abs(met[:, None] - acq_end.values).argmin(axis=1)
+    # There are 28 spins per epoch (1 aggregated science cycle)
+    # Set the spin_cycle_num to the spin number relative to the
+    # start of the ASC
+    spin_cycle_num = l1b_de["spin_cycle"] % 28
+    # Get the seconds portion of the start time for each spin
+    start_sec_spins = spin_data["start_sec_spin"].values[
+        closest_stop_acq_indices, spin_cycle_num
+    ]
+    # Get the subseconds portion of the spin start time and convert from
+    # microseconds to seconds
+    start_subsec_spins = (
+        spin_data["start_subsec_spin"].values[closest_stop_acq_indices, spin_cycle_num]
+        * 1e-6
+    )
+
+    # Combine the seconds and subseconds to get the start time for each spin
+    spin_start_time = start_sec_spins + start_subsec_spins
+    return xr.DataArray(spin_start_time)
+
+
+def set_event_met(
+    l1a_de: xr.Dataset,
+    l1b_de: xr.Dataset,
+    spin_start_time: xr.DataArray,
+    avg_spin_durations: xr.DataArray,
+) -> xr.Dataset:
+    """
+    Get the event MET for each direct event.
+
+    Each direct event is converted from a data number to engineering unit in seconds.
+    de_eu_time de_dn_time / 4096 * avg_spin_duration
+    where de_time is the direct event time Data Number (DN) and avg_spin_duration
+    is the average spin duration for the ASC that the event was measured in.
+
+    The direct event time is the time of direct event relative to the start of the spin.
+    The event MET is the sum of the start time of the spin and the
+    direct event EU time.
+
+    Parameters
+    ----------
+    l1a_de : xr.Dataset
+        The L1A DE dataset.
+    l1b_de : xr.Dataset
+        The L1B DE dataset.
+    spin_start_time : np.ndarray
+        The start time for the spin that each direct event is in.
+    avg_spin_durations : xr.DataArray
+        The average spin duration for each epoch.
+
+    Returns
+    -------
+    l1b_de : xr.Dataset
+        The L1B DE dataset with the event MET.
+    """
+    counts = l1a_de["de_count"].values
+    de_time_asc_groups = np.split(l1a_de["de_time"].values, np.cumsum(counts)[:-1])
+    de_times_eu = []
+    for i, de_time_asc in enumerate(de_time_asc_groups):
```

Comment:
I thought this looked familiar, because you are also doing this similar logic in `set_spin_cycle`. I'm wondering if it would make sense to try and combine this logic / consolidate in some way? It looks like there you are using `esa_step` as your ASC group instead of `de_time` here, so maybe it wouldn't be easy to do that though. It just struck me as similar bookkeeping.

A thought would be that rather than keeping these all as separate datasets, could you combine them into something like a "asc_datasets" that has the various dimensions and variables from the other products in it so you could align things across each one in that sense and pass in a single dataset here where you could iterate over ASC groups once and add multiple variables within that ASC loop.

Reply:
Yeah, I'd still like to implement what you suggested before with `isel` and have a ticket written to do that later. I'll add this line to that ticket and see if there's a way to combine this.